lib/3scale/backend/worker_async.rb

```diff
@@ -39,17 +39,12 @@ def work
 
         fetch_jobs_thread = start_thread_to_fetch_jobs
 
-        loop do
-          break if @shutdown
-          schedule_jobs
-          @reactor.run
-        end
+        Async { process_all }
 
         fetch_jobs_thread.join
 
         # Ensure that we do not leave any jobs in memory
-        @reactor.async { perform(@jobs.pop) } until @jobs.empty?
-        @reactor.run
+        Async { clear_queue }
```

Comment:
What is the difference between `process_all` and `clear_queue`?

Also shouldn't we wrap the whole thing inside a `Sync` call as described in the best practices link you posted?

Reply:
`process_all` is the main loop that waits forever for new jobs in the queue. When a new job is added to the queue, `@job_fetcher` pops it and adds it to `@jobs`. When interrupted, it could happen that between [worker_async.rb:76](https://github.com/3scale/apisonator/blob/a26ef8e07f052dfb1b6792fe73ebc440bbec541c/lib/3scale/backend/worker_async.rb#L76) and the shutdown of the fetcher at [worker_async.rb:44](https://github.com/3scale/apisonator/blob/a26ef8e07f052dfb1b6792fe73ebc440bbec541c/lib/3scale/backend/worker_async.rb#L44) some new jobs are added to `@jobs`. Those are already removed from the queue in Redis, so they must be processed. That's what `clear_queue` does.

Reply:
> Also shouldn't we wrap the whole thing inside a Sync call as described in the best practices link you posted?

Are you expecting me to read the links I post?

Well, using `Sync` would alter the behavior, because AFAIK falcon creates a root-level reactor, so all calls to `Async` here add tasks to the parent reactor. However, if we use `Sync`, each call would create it's own reactor that schedules the tasks under it, isolated from other tasks.

Anyway, for this particular case, I think nobody would notice any difference.

Reply:
OK, I think you mean this: https://socketry.github.io/async/guides/best-practices/index#use-a-top-level-sync-to-denote-the-root-of-your-program

Placing a `Sync` to denote the root of the program. Where do you suggest to place it? Would you wrap all the `#work` method under it?